File: oop/anemic/src/main/java/de/hhu/anemic/warenkorb/domain/model/artikel/ArtikelRepository.java
package de.hhu.anemic.warenkorb.domain.model.artikel;

import org.springframework.stereotype.Repository;

@Repository
public interface ArtikelRepository {
    Artikel findeMit(ArtikelID artikelId);
    void speichere(Artikel artikel);
    void entferne(ArtikelID artikelID);
}

File: fp/rich/src/main/clojure/de/hhu/rich/warenkorb/domain/model/artikel/ArtikelRepository.clj
(ns de.hhu.rich.warenkorb.domain.model.artikel.ArtikelRepository
  (:refer-clojure :exclude [find]))

(def find nil)
(def add! nil)
(def update! nil)

(defprotocol ArtikelRepository
  (-find [this artikel-id])
  (-add! [this artikel])
  (-update! [this concurrency-version artikel]))

(defn set-implementation! [impl]
  (def find (partial -find impl))
  (def add! (partial -add! impl))
  (def update! (partial -update! impl)))
